sound/soc/codecs/es8326.c

```diff
@@ -0,0 +1,854 @@
+
+		if (!i2c_check_functionality(adapter, I2C_FUNC_I2C)) {
+				dev_warn(&adapter->dev,
+					"I2C-Adapter doesn't support I2C_FUNC_I2C\n");
+				return -EIO;
+		}
+
+		es8326 = devm_kzalloc(&i2c->dev, sizeof(struct es8326_priv), GFP_KERNEL);
+		if (es8326 == NULL)
+				return -ENOMEM;
+
+		i2c_set_clientdata(i2c, es8326);
+		es8326->i2c = i2c;
+		mutex_init(&es8326->lock);
+		es8326->regmap = devm_regmap_init_i2c(i2c, &es8326_regmap_config);
+		if (IS_ERR(es8326->regmap)) {
+				ret = PTR_ERR(es8326->regmap);
+				dev_err(&i2c->dev, "Failed to init regmap: %d\n", ret);
+				return ret;
+		}
+
+		es8326->mclk = devm_clk_get_optional(&i2c->dev, "mclk");
+		if (IS_ERR(es8326->mclk)) {
+				dev_err(&i2c->dev,"%s,unable to get mclk\n", __func__);
+				return PTR_ERR(es8326->mclk);
+		}
+		if (!es8326->mclk)
+				dev_err(&i2c->dev,"%s, assuming static mclk\n", __func__);
+				ret = clk_prepare_enable(es8326->mclk);
+		if (ret) {
+				dev_err(&i2c->dev,"%s, unable to enable mclk\n", __func__);
+				return ret;
+		}
+
+		reg = 0x00;
+		ret = i2c_master_recv(i2c, &reg, 1);
+		if (ret < 0) {
+				dev_err(&i2c->dev, "i2c recv Failed\n");
+				return ret;
+		}
+		es8326->irq = i2c->irq;
+		ret = devm_request_threaded_irq(&i2c->dev, es8326->irq, NULL, es8326_irq,
+									IRQF_TRIGGER_RISING | IRQF_ONESHOT,
+									"es8326", es8326);
+		if(ret == 0)	
+		{   
+				disable_irq(es8326->irq);
+		}
+		else{
+				dev_warn(&i2c->dev, "Getting irq failed.");
+		}
+		
+		ret = of_property_read_u8(i2c->dev.of_node, "amic1-src", &es8326->amic1_src);
+		if (ret != 0) {
+				dev_dbg(&i2c->dev, "amic1-src return %d", ret);
+				es8326->amic1_src = ES8326_MIC1_SEL;
+		}
+
+		ret = of_property_read_u8(i2c->dev.of_node, "amic2-src", &es8326->amic2_src);
+		if (ret != 0) {
+				dev_dbg(&i2c->dev, "amic2-src return %d", ret);
+				es8326->amic2_src = ES8326_MIC1_SEL;
+		}
+
+		ret = of_property_read_u8(i2c->dev.of_node, "mic1-src", &es8326->mic1_src);
+		if (ret != 0) {
+				dev_dbg(&i2c->dev, "mic1-src return %d", ret);
+				es8326->mic1_src = ES8326_ADC_AMIC;
+		}
+		dev_dbg(&i2c->dev, "mic1-src %x",es8326->mic1_src);
+
+		ret = of_property_read_u8(i2c->dev.of_node, "mic2-src", &es8326->mic2_src);
+		if (ret != 0) {
+				dev_dbg(&i2c->dev, "mic2-src return %d", ret);
+				es8326->mic2_src = ES8326_ADC_DMIC;
+		}
+		dev_dbg(&i2c->dev, "mic2-src %x",es8326->mic2_src);
+		
+		ret = of_property_read_u8(i2c->dev.of_node, "jack-pol", &es8326->jack_pol);
+		if (ret != 0) {
+				dev_dbg(&i2c->dev, "jack-pol return %d", ret);
+				es8326->jack_pol = ES8326_HP_DET_JACK_POL|ES8326_HP_DET_BUTTON_POL|ES8326_HP_TYPE_OMTP;
+		}
+		dev_dbg(&i2c->dev, "jack-pol %x",es8326->jack_pol);
+
+		ret = of_property_read_u32(i2c->dev.of_node, "mclk-rate", &es8326->mclk_rate);
+		if (ret != 0) {
+				dev_dbg(&i2c->dev, "mclk-rate return %d", ret);
+				es8326->mclk_rate = 19200000;
+		}
+		dev_dbg(&i2c->dev, "mclk-rate %u",es8326->mclk_rate);
```

Comment:
what about jack detect inversion, is this not supported?

Reply:
defined as ES8326_HP_DET_JACK_POL. Is a device properity named jd_inverted necessary?

Reply:
well, if you want to use this codec driver with the same machine driver, then the device properties and quirks have to be supported for all codec drivers. I don't want to special case this es8326 driver for property handling.